python/revng/model/base.py

```diff
@@ -0,0 +1,170 @@
+from dataclasses import dataclass, fields
+from enum import Enum
+from typing import Generic, Type, TypeVar, get_args, get_origin, get_type_hints
+
+import yaml
+
+
+no_default = object()
+
+
+@dataclass
+class StructBase:
+    @classmethod
+    def from_dict(cls, **kwargs):
+        constructor_kwargs = {}
+
+        for field_name, field_value in kwargs.items():
+            field_spec = cls.__dataclass_fields__.get(field_name)
+            if field_spec is None:
+                raise ValueError(f"Field {field_name} is not allowed for type {cls.__name__}")
+
+            try:
+                # isinstance raises if field_spec.type is an instance of `typing._SpecialForm`
+                # such as `Optional[something]`
+                if isinstance(field_value, field_spec.type):
+                    constructor_kwargs[field_name] = field_value
+                    continue
+            except TypeError:
+                pass
+
+            if isinstance(field_spec.type, str):
+                field_spec_type = get_type_hints(cls)[field_spec.name]
+            else:
+                field_spec_type = field_spec.type
+
+            origin = get_origin(field_spec_type)
+            args = get_args(field_spec_type)
+
+            if origin is list:
+                assert len(args) == 1
+                underlying_type = args[0]
+
+                if not isinstance(field_value, list):
+                    raise ValueError(f"Expected list for field {field_name} of {cls.__name__}, got {type(field_value)} instead")
+
+                instances = []
+                for v in field_value:
+                    if isinstance(v, underlying_type):
+                        v_inst = v
+                    elif isinstance(v, str) and issubclass(underlying_type, Enum):
+                        v_inst = underlying_type(v)
+                    elif isinstance(v, dict) and issubclass(underlying_type, StructBase):
+                        v_inst = underlying_type.from_dict(**v)
+                    elif isinstance(v, str) and hasattr(underlying_type, "from_string"):
+                        v_inst = underlying_type.from_string(v)
+                    else:
+                        raise ValueError(f"Unhandled value of type {type(v)}: {v}")
```

Comment:
It feels like this should be its own function: it's deserialization logic even more generic than `StructBase.from_dict`.